src/OpenTelemetry/Logs/OpenTelemetryLogger.cs

```diff
@@ -115,12 +147,65 @@ public bool IsEnabled(LogLevel logLevel)
                     return attributeStorage;
                 }
             }
+            else if (!parseStateValues || state is null)
+            {
+                // Note: This is to preserve legacy behavior where State is
+                // exposed if we didn't parse state.
+                iLoggerData.State = state;
+                return null;
+            }
             else
             {
-                var attributeStorage = logRecord.AttributeStorage ??= new List<KeyValuePair<string, object?>>(LogRecordPoolHelper.DefaultMaxNumberOfAttributes);
-                attributeStorage.Add(new KeyValuePair<string, object?>(string.Empty, state));
-                return attributeStorage;
+                try
+                {
+                    PropertyDescriptorCollection itemProperties = TypeDescriptor.GetProperties(state);
```

Comment:
Why do we want to add this new feature of parsing through custom typed objects?

Reply:
To make them useful, basically. Today these logs get a `StateValues` of `[string.Empty] = state`. Probably most exporters are just dropping that? Some might do a `ToString` on it. This change is to fulfill the promise of what `ParseStateValues` was intended to do which is parse these things into something useful!

Reply:
This could lead to exporting not so useful data in some cases. For example:

```csharp
logger.Log(
    LogLevel.Information,
    eventId: default,
    "Test",
    exception: null,
    formatter: null);
```

This log statement would have had the `stateValues` as `[string.Empty] = "Test"` but now we would have the `StateValues` as `"Length" = 4` which is the `Length` property of the string.

Similarly, for `DateTime` we would now log all the [properties ](https://learn.microsoft.com/en-us/dotnet/api/system.datetime?view=net-8.0#properties) of `DateTime` instead of simply exporting `[string.Empty] = dateTime.ToString()`.

Reply:
This is a good point that I agree with but not sure if it is a concern. I feel like if you are logging something simple like a string or DateTime the extensions work nicely and would never hit this code. Log<T>() seems geared for more advanced/complicated things? `ParseStateValues` is going to parse the state into values so you get what you ask for :rofl:

If we do this I do expect some users to ask for the ability to take over the serialization. I didn't tackle that here but I could see allowing them to own it via some service registration 🤷